torch/cuda/__init__.py

```diff
@@ -500,15 +500,18 @@ def _raw_device_count_nvml() -> int:
     return dev_arr[0]
 
 def _device_count_nvml() -> int:
+    visible_devices = _parse_visible_devices()
+    if len(visible_devices) == 0:
+        return 0
     try:
         raw_cnt = _raw_device_count_nvml()
-        if raw_cnt <= 0:
-            return raw_cnt
-        return len(set(range(raw_cnt)).intersection(_parse_visible_devices()))
     except OSError:
         return -1
     except AttributeError:
         return -1
+    if raw_cnt <= 0:
+        return raw_cnt
```

Comment:
what is semantics of device count == -1? currently the documentation just says `Returns the number of GPUs available.`?

Reply:
If `-1` is not a special value, then maybe `0` should be returned in this case as well?

Reply:
`device_count` should never return `-1`, but  `_device_count_nvml()`  returning negative value, means that it failed to query number of devices via NVML and system will fall back to old path using CUDA runtime API.

And even though function is internal, let me add a line documenting this behavior (which is a very common of POSIX, but quite unPythonic)

Reply:
If the return value of the internal function gets at the end forwarded / returned further to the user, it may be a problem

Reply:
Documented, although it is an internal function.

Reply:
essentially, it works because set(range(-1)) is an empty set?